localstack/services/cloudformation/engine/template_preparer.py

```diff
@@ -29,21 +37,156 @@ def template_to_json(template: str) -> str:
+                parsed_template=result,
+                macro=transformation,
+                stack_parameters=parameters,
+            )
+
+    return result
+
+
+def execute_macro(parsed_template: Dict, macro: Dict, stack_parameters: List) -> str:
+    macro_definition = get_cloudformation_store().macros.get(macro["Name"])
+    if not macro_definition:
+        raise FailedTransformation(macro["Name"], "2DO")
+
+    formatted_stack_parameters = {
+        param["ParameterKey"]: param["ParameterValue"] for param in stack_parameters
+    }
+
+    formatted_transform_parameters = macro.get("Parameters", {})
+    for k, v in formatted_transform_parameters.items():
+        if isinstance(v, Dict) and "Ref" in v:
+            formatted_transform_parameters[k] = formatted_stack_parameters[v["Ref"]]
+
+    transformation_id = f"{get_aws_account_id()}::{macro['Name']}"
+    event = {
+        "region": aws_stack.get_region(),
+        "accountId": get_aws_account_id(),
+        "fragment": parsed_template,
+        "transformId": transformation_id,
+        "params": formatted_transform_parameters,
+        "requestId": long_uid(),
+        "templateParameterValues": formatted_stack_parameters,
+    }
+
+    client = aws_stack.connect_to_service("lambda")
+    invocation = client.invoke(
+        FunctionName=macro_definition["FunctionName"], Payload=json.dumps(event)
+    )
+    if invocation.get("StatusCode") != 200 or invocation.get("FunctionError") == "Unhandled":
+        raise FailedTransformation(
+            transformation=macro["Name"],
+            message=f"Received malformed response from transform {transformation_id}. Rollback requested by user.",
+        )
+    result = json.loads(invocation["Payload"].read())
+
+    if result.get("status") != "success":
+        error_message = result.get("errorMessage")
+        message = (
+            f"Transform {transformation_id} failed with: {error_message}. Rollback requested by user."
+            if error_message
+            else f"Transform {transformation_id} failed without an error message.. Rollback requested by user."
+        )
+        raise FailedTransformation(transformation=macro["Name"], message=message)
+
+    if not isinstance(result.get("fragment"), dict):
+        raise FailedTransformation(
+            transformation=macro["Name"],
+            message="Template format error: unsupported structure.. Rollback requested by user.",
+        )
+
+    return result.get("fragment")
+
+
+def apply_serverless_transformation(parsed_template):
     """only returns string when parsing SAM template, otherwise None"""
-    # template_body = get_template_body(req_data) # FIXME
-    parsed = parse_template(template_body)
-    if parsed.get("Transform") == "AWS::Serverless-2016-10-31":
+    region_before = os.environ.get("AWS_DEFAULT_REGION")
+    if boto3.session.Session().region_name is None:
+        os.environ["AWS_DEFAULT_REGION"] = aws_stack.get_region()
```

Comment:
Is this something that happened to you?

I'd heavily discourage setting any environment variables here.